$NetBSD: patch-bq,v 1.1 2011/06/06 15:53:14 manu Exp $

Make sur glusterfsd uses the same address family as gluserd. 

glusterfsd uses getaddrinfo and picks the first entry for its
listening address. The first entru is of inet7 family on 
NetBSD. Even if glusterd was started with option 
transport.address-family inet, we ended with glusterfsd 
started with inet6, and the client could not connect to it
if the letwork lack IPv6

--- xlators/mgmt/glusterd/src/glusterd-utils.c.orig	2011-06-06 10:28:42.000000000 +0200
+++ xlators/mgmt/glusterd/src/glusterd-utils.c	2011-06-06 16:17:30.000000000 +0200
@@ -88,8 +88,52 @@
         //TODO: set timestamp
         return 0;
 }
 
+static const char *
+glusterd_listener_family_name(void)
+{
+	xlator_t	 	*this = NULL;
+	glusterd_conf_t		*conf = NULL;
+	rpcsvc_listener_t	*listener = NULL;
+	struct sockaddr_storage ss;
+
+	this = THIS;
+	GF_ASSERT (this);
+	conf = this->private;
+	GF_ASSERT (conf);   
+	GF_ASSERT (conf->rpc);
+	
+	list_for_each_entry (listener, &conf->rpc->listeners, list)
+		break; /* grab first one */
+
+	GF_ASSERT (listener->trans);
+
+	if (rpc_transport_get_myaddr(listener->trans, NULL, 0, 
+				     &ss, sizeof(ss)) != 0) {
+		gf_log ("glusterd", GF_LOG_ERROR,
+			"rpc_transport_get_myname failed: %s",
+			strerror(errno));
+		return NULL;
+	}
+
+	switch (ss.ss_family) {
+		case AF_INET:
+			return "inet";
+			break;
+		case AF_INET6:
+			return "inet6";
+			break;
+		default:
+			gf_log ("glusterd", GF_LOG_ERROR,
+				"unknown address family %d",
+				ss.ss_family);
+			break;
+	}
+	       
+	return NULL;
+}
+
 gf_boolean_t
 glusterd_is_loopback_localhost (const struct sockaddr *sa, char *hostname)
 {
         GF_ASSERT (sa);
@@ -1026,8 +1070,10 @@
         int                     port = 0;
         FILE                    *file = NULL;
         gf_boolean_t            is_locked = _gf_false;
         char                    socketpath[PATH_MAX] = {0};
+	const char		*family_name;
+        char                    *family_option[8192] = {0,};
 
         GF_ASSERT (volinfo);
         GF_ASSERT (brickinfo);
 
@@ -1106,14 +1152,19 @@
         port = brickinfo->port;
         if (!port)
                 port = pmap_registry_alloc (THIS);
 
+	if ((family_name = glusterd_listener_family_name()) != NULL) 
+		snprintf (family_option, 8192, 
+			  "--xlator-option %s-server.transport.address-family=%s ",
+			  volinfo->volname, family_name);
+
         snprintf (cmd_str, 8192,
                   "%s/sbin/glusterfsd --xlator-option %s-server.listen-port=%d "
-                  "-s localhost --volfile-id %s -p %s -S %s --brick-name %s "
-                  "--brick-port %d -l %s", GFS_PREFIX, volinfo->volname,
-                  port, volfile, pidfile, socketpath, brickinfo->path, port,
-                  brickinfo->logfile);
+                  "%s -s localhost --volfile-id %s -p %s -S %s --brick-name %s "
+                  "--brick-port %d -l %s", GFS_PREFIX, volinfo->volname, port, 
+		  family_option, volfile, pidfile, socketpath, brickinfo->path, 
+		  port, brickinfo->logfile);
 
 	gf_log ("",GF_LOG_DEBUG,"Starting GlusterFS Command Executed: \n %s \n", cmd_str);
         ret = gf_system (cmd_str);
 
